### DISCUSSION TITLE
Blueman not displaying code transmitted by keyboard when trying to connect

### DISCUSSION THREAD

**Discussion by aussiedaniel** [blueman-project/blueman#2167] [Q&A]
blueman:
BlueZ:
Distribution:
Desktop environment:

->When trying to connect my bluetooth keyboard blueman asks for the code. I don't have the code. The code is a random number transmitted by the keyboard that needs to be displayed by blueman. Once this code is displayed it is entered on the bluetooth keyboard to connect.

I submitted a defect on bugzilla but the assigned engineer didn't have access to a bluetooth keyboard and suggested I should report the issue upstream.  The following is a link to this bugzilla defect report.
https://bugzilla.redhat.com/show_bug.cgi?id=2192091

**Comment by infirit** [comment]
Blueman will create a notification with the number. Which is actually in the log you posted on the rhbz.

**Reply by aussiedaniel** [reply]
The log I submitted of the number being displayed was from the now unsupported product blueberry. I submitted this log to show what should happen in blueman.

**Reply by infirit** [reply]
This log, https://bugzilla-attachments.redhat.com/attachment.cgi?id=1961437.

`BluezAgent:164 _on_request_pin_code: Agent.RequestPinCode` will create a notification with pin to be displayed by whatever notification deamon you are using. 7 seconds later it expires `Notification:248 do_g_signal: The notification expired.`.

Blueman relies on a proper notification deamon running.

**Reply by aussiedaniel** [reply]
The mate-notification-daemon only runs briefly to display the message. There is a test of the notification daemon inside the mate control control panel. The test was successful.

Do have access to a bluetooth keyboard to test? If the bluetooth keyboard works correctly on your system, then it is just a matter of finding what is different between your system and my system.

**Reply by infirit** [reply]
It shows the notification fine on plasma including the number to be entered in bold. Can you pair it through the commandine with `bluetoothctl`.

You mention mate notification daemon, you running MATE as desktop?

**Reply by aussiedaniel** [reply]
Yes, I'm running MATE as desktop

Here is the output from bluetoothctl:
[bluetooth]# scan on
Discovery started
[CHG] Controller D0:AB:D5:93:52:46 Discovering: yes
[NEW] Device 58:71:33:66:7B:97 Bluetooth Keyboard
[bluetooth]# connect 58:71:33:66:7B:97
Attempting to connect to 58:71:33:66:7B:97
[CHG] Device 58:71:33:66:7B:97 Connected: yes
[CHG] Device 58:71:33:66:7B:97 Modalias: bluetooth:v04E8p7021d0100
[CHG] Device 58:71:33:66:7B:97 UUIDs: 00001124-0000-1000-8000-00805f9b34fb
[CHG] Device 58:71:33:66:7B:97 UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[CHG] Device 58:71:33:66:7B:97 ServicesResolved: yes
[CHG] Device 58:71:33:66:7B:97 WakeAllowed: yes
Connection successful
Request PIN code
[agent] Enter PIN code: [DEL] Device 58:71:33:66:7B:97 Bluetooth Keyboard
[bluetooth]# 

It asking for the code, but no code was displayed from the keyboard
It then times out and deletes the keyboard

**Comment by infirit** [comment]
> Yes, I'm running MATE as desktop
> 
> Here is the output from bluetoothctl: 

```
[bluetooth]# scan on
Discovery started
[CHG] Controller D0:AB:D5:93:52:46 Discovering: yes
[NEW] Device 58:71:33:66:7B:97 Bluetooth Keyboard
[bluetooth]# connect 58:71:33:66:7B:97
Attempting to connect to 58:71:33:66:7B:97
[CHG] Device 58:71:33:66:7B:97 Connected: yes
[CHG] Device 58:71:33:66:7B:97 Modalias: bluetooth:v04E8p7021d0100
[CHG] Device 58:71:33:66:7B:97 UUIDs: 00001124-0000-1000-8000-00805f9b34fb
[CHG] Device 58:71:33:66:7B:97 UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[CHG] Device 58:71:33:66:7B:97 ServicesResolved: yes
[CHG] Device 58:71:33:66:7B:97 WakeAllowed: yes
Connection successful
Request PIN code
[agent] Enter PIN code: [DEL] Device 58:71:33:66:7B:97 Bluetooth Keyboard
```
> 
> It asking for the code, but no code was displayed from the keyboard It then times out and deletes the keyboard

Apologies, RequestPinCode does not display a pincode. I was thinking of DisplayPinCode. What should happen is that a dialog is opened where the user, you @aussiedaniel , need to enter a pin code. Does the applet show you a dialog? It may be hidden under other windows.

**Reply by aussiedaniel** [reply]
I minimized all windows so that I wouldn't miss a popup displaying the code from the keyboard, but there was no popup.

I also tried this on my laptop which is completely different hardware. On the laptop a popup came up asking for the code, but no code from the keyboard was displayed. I also had all windows minimized so I won't miss a popup displaying the code from the keyboard.

**Reply by infirit** [reply]
I have to dig through my keyboards and see if I can find one that goes through RequestPinCode.

Which keyboard are you using btw?

**Reply by cschramm** [reply]
I would not expect `RequestPinCode` to be called for a keyboard at all. Given that our agent has the `KeyboardDisplay` capability, BlueZ calls it for legacy pairing a device with a display capability. That obviously does not make sense for an ordinary keyboard.

Anyway, as the bluetoothctl agent shows exactly the same, blueman's agent is not at fault here, but there's a more generic issue with the keyboard or something. Did you consult its manual for the intended pairing sequence?

**Reply by aussiedaniel** [reply]
I have two bluetooth keyboards and they both act the same way with blueman. The older keyboard was purchased about 5 years ago and the newer keyboard I bought in 2023. So the newer keyboard is not that old. I will attach the manual for the newer keyboard. The manual for the older keyboard is lost. Both keyboards work as expected with the now obsolete product blueberry. Also both keyboards work as expected with my  Lenovo tablet. Maybe you could check the blueberry code for something special about keyboards.
[manual.pdf](https://github.com/blueman-project/blueman/files/13114048/manual.pdf)

**Reply by cschramm** [reply]
According to that manual I'd not expect BlueZ to call `RequestPinCode` / `RequestPasskey` but `DisplayPinCode` / `DisplayPasskey` (`PinCode` is legacy pairing, `Passkey` is SSP). Note that anything not older than 15 years should support SSP, so if `PinCode` stuff shows up, either your local or the remote device's chip is something old / cheap / counterfeit / strange / whatever.

Regarding the difference between blueman / bluetoothctl and blueberry: blueberry uses gnome-bluetooth and gnome-bluetooth registers its agent with capability `DisplayYesNo`. You can try `bluetoothctl --agent DisplayYesNo` to see if that gives you the blueberry behavior.

**Reply by cschramm** [reply]
I'd be keen to see a btmon trace of the process that leads to the `RequestPinCode` call.

**Reply by infirit** [reply]
This is the process on my Logitech MX5000 keyboard. It's an ancient 15+ years old bluetooth 2.0 keyboard.
1. Connect
2. Dialog shows to enter pin (see below)
3. Enter same pin on keyboard
4. Paired properly and working on plasma

blueman-applet log for this part
```
blueman-applet 19.54.24 DEBUG    Base:67 _properties_changed: /org/bluez/hci0/dev_00_07_61_3D_74_E7 {'UUIDs': ['00001124-0000-1000-8000-00805f9b34fb', '00001200-0000-1000-8000-00805f9b34fb'], 'ServicesResolved': True, 'WakeAllowed': True} [] <Device.Device object at 0x7feaf791a980 (blueman+bluez+Device+Device at 0x5584e9f2b230)>
blueman-applet 19.54.25 INFO     BluezAgent:164 _on_request_pin_code: Agent.RequestPinCode
blueman-applet 19.54.28 INFO     Notification:242 do_g_signal: 1
blueman-applet 19.54.28 DEBUG    Notification:246 do_g_signal: The notification expired.
blueman-applet 19.55.05 DEBUG    Base:67 _properties_changed: /org/bluez/hci0/dev_00_07_61_3D_74_E7 {'Bonded': True, 'Paired': True} [] <Device.Device object at 0x7feaf791a980 (blueman+bluez+Device+Device at 0x5584e9f2b230)>
```
![image](https://github.com/blueman-project/blueman/assets/3465730/cc9a64cd-c578-4625-9aab-3be0c23c0fdd)

**Reply by cschramm** [reply]
Seems like a generic issue then. :eyes: I'm totally surprised to see that coming up.

A little background on the capability: gnome-bluetooth basically always had `DisplayYesNo` for its agent, adopted from gnome-bluez nearly 15 years ago. I fail to find any archive for gnome-bluez, so I'm left with no clue about the reason. It does not make much sense to me as while one might argue that the system might not have a keyboard connected (yet), it might not have a display either and I do not get what the "yes / no" input capability should be (a power button? :laughing:). blueman 1 always had `DisplayYesNo` as well. I guess that it also was just taken from gnome-bluez. [It came up in blueman 1.99.alpha1](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=780776) that when some peripheral - presumably with `DisplayOnly` capability - got paired (using SSP) while disallowing unauthenticated connections, BlueZ (5.23) just gave an error if the agent did not have keyboard capability. That's why blueman 2 always used `DisplayKeyboard`, making much more sense to me as we can assume the system to be capable to enter a code.

If using `DisplayKeyboard` generally leads to the very unexpected behavior with Bluetooth 2.0 keyboards described here (i.e. a request to enter an unknown pin code instead of displaying one), I wonder how that never came up in the last 8 years of blueman 2 (well, we already even had Bluetooth 4.2 all the time, but obviously 2.0 keyboards are still a thing).

I'm still interested in a btmon trace to understand the unexpected pin code request (maybe a debug log from bluetoothd could also shine some light). Also, unless BlueZ changed something (I cannot reckon what, though), the Debian bug for blueman should still be reproducible with gnome-bluetooth today.

**Reply by cschramm** [reply]
Oh, I get it. With 2.0 legacy pairing you always had to enter the same arbitrary PIN code on both ends (one end might hardcode it, so that you have to type the hardcoded one on the other end). That's what we're seeing.

However, when disabling the keyboard capability on our end, BlueZ will also downgrade `Passkey Entry` for SSP to `Just works`. :confused:

**Reply by infirit** [reply]
Yes, this ancient keyboard has the legacy pairing property set to true. I'll look through my keyboards if I can find a newer one.

**Reply by aussiedaniel** [reply]
I would be happy to supply a btmon trace, but you will have to give me instructions on how to obtain one.

**Reply by cschramm** [reply]
Just `sudo btmon -w /tmp/trace` until you get the pin code request, but I guess I got what's happening and will wrap my head around some BlueZ code instead.

By the way: Try 0000 for the PIN code request. It might just work. :wink:

**Reply by infirit** [reply]
~~I found a slightly later model, mx5500 which has legacy pairing false but BlueZ calls RequestPinCode and I have to provide pin just like before. I don't have any bluetooth 3 boards I think. `bluetoothctl --agent DisplayYesNo` makes not difference for these.~~ It's legacy pairing now that it paired/connected, I swear it was false when I first tried it :thinking: 

```
blueman-applet 18.24.30 INFO     RecentConns:112 notify    : <Device.Device object at 0x7fdd052bbe80 (blueman+bluez+Device+Device at 0x55e4c1126510)> 00000000-0000-0000-0000-000000000000
blueman-applet 18.24.31 DEBUG    Base:67 _properties_changed: /org/bluez/hci0/dev_00_1F_20_26_7E_44 {'Connected': True} [] <Device.Device object at 0x7fdd052bbe80 (blueman+bluez+Device+Device at 0x55e4c1126510)>
blueman-applet 18.24.32 DEBUG    Base:67 _properties_changed: /org/bluez/hci0/dev_00_1F_20_26_7E_44 {'Modalias': 'usb:v046DpB30Bd0100'} [] <Device.Device object at 0x7fdd052bbe80 (blueman+bluez+Device+Device at 0x55e4c1126510)>
blueman-applet 18.24.32 DEBUG    Base:67 _properties_changed: /org/bluez/hci0/dev_00_1F_20_26_7E_44 {'UUIDs': ['00001124-0000-1000-8000-00805f9b34fb', '00001200-0000-1000-8000-00805f9b34fb'], 'ServicesResolved': True, 'WakeAllowed': True} [] <Device.Device object at 0x7fdd052bbe80 (blueman+bluez+Device+Device at 0x55e4c1126510)>
blueman-applet 18.24.32 INFO     BluezAgent:164 _on_request_pin_code: Agent.RequestPinCode
blueman-applet 18.24.36 INFO     Notification:242 do_g_signal: 1
blueman-applet 18.24.36 DEBUG    Notification:246 do_g_signal: The notification expired.
blueman-applet 18.24.46 DEBUG    Base:67 _properties_changed: /org/bluez/hci0/dev_00_1F_20_26_7E_44 {'Bonded': True, 'Paired': True} [] <Device.Device object at 0x7fdd052bbe80 (blueman+bluez+Device+Device at 0x55e4c1126510)>
```

**Reply by cschramm** [reply]
Alright, so...

Bluetooth 2.0 (and older) generally involves a PIN code request on both ends. That's what we see. If you enter a random PIN code in blueman (using an already connected keyboard) and on the Bluetooth keyboard, it should work as intended.

On top of that, the HID specification recommends that when the remote device is a keyboard, the local device should just provide a random PIN code in response to the request and display it to the user for entering the same code on the keyboard.

bluetoothd receives the request as `MGMT_EV_PIN_CODE_REQUEST` from the kernel. Opposed to Secure Simple Pairing, where it receives distinct events for requesting and displaying a passkey, it decides to either call the agent's `RequestPinCode` or `DisplayPinCode` in response to the same kernel event.

The most important callback to find a PIN code and call `DisplayPinCode` instead of `RequestPinCode` is BlueZ' autopair plugin and that - apart from other magic - implements exactly the recommendation to generate a random code if the remote device is a keyboard. I'd expect that to happen completely independent of the agent.

Nowhere in that process does bluetoothd check the agent's capabilities, so that I would neither expect it to force `RequestPinCode` if keyboard capability is provided, nor to suppress it otherwise. That would mean that the agent capability is *not* the crucial difference between blueman and gnome-bluetooth / blueberry and that brings me to...

gnome-bluetooth has what it calls a PIN code database, doing similar things like autopair and also including a special keyboard handling. That actually generates a random PIN code for keyboards and *that* is the crucial difference between the agents.

We dropped the PIN database from blueman in 2.2.4 (#1614) as our implementation was broken and seemed obsolete given BlueZ' autopair plugin. That could be the reason why the topic did not come up for most of blueman 2's lifetime, *but* one thing that was broken is that the implementation most probably always provided the string `KEYBOARD` as a PIN code for keyboards, so that I wonder if that ever worked.

However, the more important question is why the autopair plugin does not seem to work for you. Maybe a bluetoothd debug log could tell us.

**Reply by aussiedaniel** [reply]
I would be happy to provide a bluetoothd debug log, but you have to give me instructions on how to obtain one.

**Reply by cschramm** [reply]
Not sure if there are any Fedora specifics, but generally you should be able to `sudo systemctl stop bluetooth.service` and then `sudo /usr/libexec/bluetooth/bluetoothd -d -n` to run the daemon with debug logging in foreground. Just copy the output. To revert, ctrl-c the process and `sudo systemctl start bluetooth.service`.

(The Fedora wiki recommends a different way where you edit the service and fetch its logs from `journalctl` in case that does not work: https://fedoraproject.org/wiki/How_to_debug_Bluetooth_problems#Simple_debugging)

**Reply by infirit** [reply]
Here is the log for the mx5500 https://gist.github.com/infirit/940e4db2a5ce6caaa5c585b28e6de86a. I do have bluetooth3 keyboard, the k480 which just connects without any pin request.

**Reply by infirit** [reply]
Perhaps an interesting observation. There is different behaviour with the mx5500, when doing generic connect and when I specifically ask to bond/pair. When bonding/pairing DisplayPinCode is actually called. Just connect goes through the dialog and asks me to provide a pin.

**Reply by aussiedaniel** [reply]
I had to use the Fedora method to get the log.  I don't know if this log is useful as I tried three times and it never asked for the pin. It just times out.
[bluetoothd.txt](https://github.com/blueman-project/blueman/files/13185637/bluetoothd.txt)

**Reply by cschramm** [reply]
I think we have the reason: BlueZ only calls PIN callbacks if there is an existing `bonding_req` for the device. That's the case when you try to pair it, but not when you try to connect to an unpaired device. So you should get a PIN code displayed when explicitly pairing but you always will be requested to enter one when connecting.

~I will forward that to BlueZ, but am still looking into the consequences.~

**Reply by infirit** [reply]
blueberry calls `GnomeBluetooth().Client.connect_service` which ends up calling the `Connect` method on a `Device1` dbus proxy. Blueman does the same with the (generic) Connect menu entry so :shrug: ?

@aussiedaniel I see you also have gnomebluetooth running with an agent. Blueman and gnomebluetooth don't coexist well so make sure gnomebluetooth isn't running (or better uninstall it). Also try the `Pair` menu entry instead of `Connect`, see how that works.

**Reply by aussiedaniel** [reply]
I tried the Pair menu entry and I get multiple connect/disconnect notifications. I have attached the log.
I aborted the blueberry agent.
 
[bluetoothd2.txt](https://github.com/blueman-project/blueman/files/13188667/bluetoothd2.txt)

**Reply by aussiedaniel** [reply]
I noticed a new bluez update and I thought I would try blueman again. When I double click on the keyboard I get a notification the keyboard connected, but the keyboard doesn't work. If I pair the device from the drop down menu I get a notification with a 6 digit number. Upon entering the 6 digit number into the keyboard and pressing enter, I get a disconnect notification. If I click connect from the drop down menu, I get a connect notification and I have a Working keyboard. This is a great improvement.
I think I should have a working keyboard as soon as I enter the 6 digit number into the keyboard and press enter.  I also think the 6 digit number should be displayed when I double click the keyboard.

**Reply by cschramm** [reply]
Quick wrap up: I opened #2178 for this. The underlying problem is that we dropped blueman's broken PIN database in favor of BlueZ' autopair plugin, but it turns out that the plugin only comes into action when the pairing is initiated locally. When you try to connect to the un-paired device, the remote device initiates the pairing. gnome-bluetooth still has its PIN database. I'd love to see a more generic autopair in BlueZ so that PIN databases do not get duplicated everywhere, but unfortunately nobody stepped up to discuss that yet (I might give it another try at some point when I find more time).

I actually thought about posting to #2178 as [the recent issues due to a changed default setting upstream and in distributions](https://github.com/bluez/bluez/issues/673) seem to affect pretty much the same devices. It's exactly what you just described.

**Comment by infirit** [comment]
> On top of that, the HID specification recommends that when the remote device is a keyboard, the local device should just provide a random PIN code in response to the request and display it to the user for entering the same code on the keyboard.

For RequestPinCode/PassCode we can check for keyboards with `get_minor_class(device["Class"]) == _("Keyboard")` and generate a 6 digit number with `secrets.SystemRandom().randint(100000, 999999)`. This for sure will make things a lot better for ancient 2.0 keyboards and other uses like Daniel.

I'll send a draft PR in the next couple of days?

**Reply by cschramm** [reply]
But that's exactly what [autopair](https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/plugins/autopair.c) does (since BlueZ 5.55). Plus, it tries 0000 next if the random 6-digit PIN gets rejected immediately.

And while this discussion leaves a different impression, it definitely does work in general, otherwise we'd basically never see `DisplayPinCode` at all.
